File: 8.bellman_ford/cpp/src/main.cpp
#include <iostream>
#include <vector>
#include <climits>

const int UNKNOWN = INT_MAX;

class Edge {
public:
    int from;
    int to;
    int weight;

    Edge(int from, int to, int weight) {
        this->from = from;
        this->to = to;
        this->weight = weight;
    }
};

int main(int argc, char* argv[]) {
    int V, E;

    FILE* file = std::fopen(argv[1], "r");
    std::fscanf(file, "%d %d", &V, &E);

    std::vector<Edge*> edges;
    std::vector<int> dist;
    for(int i=0;i<V;i++) {
        if(i==0) dist.push_back(0);
        else dist.push_back(UNKNOWN);
    }

    for(int i=0; i<E; i++) {
        int from, to, weight;
        std::fscanf(file, "%d %d %d", &from, &to, &weight);

        edges.push_back(new Edge(from, to, weight));
    }

    // bellman-ford
    for(int i=0;i<V-1;i++) {
        bool changed = false;
        for(Edge* edge: edges) {
            if(dist[edge->from] != UNKNOWN && dist[edge->to] > dist[edge->from] + edge->weight) {
                dist[edge->to] = dist[edge->from] + edge->weight;
                changed = true;
            }
        }

        if(!changed) break;
    }

    // check negative cycle
    bool has_neg_cycle = false;
    for(Edge* edge: edges) {
        if(dist[edge->to] > dist[edge->from] + edge->weight) {
            has_neg_cycle = true;
            break;
        }
    }

    if(has_neg_cycle) std::printf("Has negative cycle !!\n");
    else {
        int index = 0;
        for(int d: dist) {
            printf("%d: %d\n", index, d);
            index++;
        }
    }

    return 0;
}

File: 8.bellman_ford/python/main.py
import sys
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--input_file")
args = parser.parse_args()

UNKNOWN = sys.maxsize

class Edge:
    def __init__(self, from_: int, to_: int, weight: int):
        self.from_: int = from_
        self.to_: int = to_
        self.weight: int = weight

if __name__ == "__main__":
    with open(args.input_file, "r") as f:
        V, E = tuple(map(lambda x: int(x), f.readline().split(" ")))
        
        nodes = [0 if i == 0 else UNKNOWN for i in range(V)]
        edges = []

        for line in f:
            from_, to_, weight = tuple(map(lambda x: int(x), line.split(" ")))
            edges.append(Edge(from_=from_, to_=to_, weight=weight))
    
    # bellman-ford
    for _ in range(V-1):
        changed = False
        for edge in edges:
            if nodes[edge.from_] != UNKNOWN and nodes[edge.to_] > nodes[edge.from_] + edge.weight:
                nodes[edge.to_] = nodes[edge.from_] + edge.weight
                changed = True
        
        if not changed:
            break
    
    # check for negative weighted cycle
    has_negative_cycle = False
    for edge in edges:
        if nodes[edge.to_] > nodes[edge.from_] + edge.weight:
            has_negative_cycle = True
            break
    
    if has_negative_cycle:
        print("has negative cycle")
    else:
        for idx, dist in enumerate(nodes):
            print("{} : {}".format(idx, dist))
